File: scm-csharp/scheme/SchemeHttpRequest.cs
using System.Collections.Generic;

namespace scheme;

public class SchemeHttpRequest
{
    // Default request timeout in seconds. A value <= 0 means "no timeout".
    public const int DefaultTimeoutSeconds = 600;

    public string Method;
    public string Url;
    public List<(string Name, string Value)> Headers;
    public string? Body;        // UTF-8 view of BodyBytes (for back-compat).
    public byte[]? BodyBytes;   // null if no body — the raw, byte-correct body.
    public int TimeoutSeconds = DefaultTimeoutSeconds; // <= 0 means no timeout

    public SchemeHttpRequest(string method, string url, List<(string, string)> headers, string? body)
        : this(method, url, headers, body, null)
    {
    }

    public SchemeHttpRequest(string method, string url, List<(string, string)> headers,
                             string? body, byte[]? bodyBytes)
    {
        Method = method;
        Url = url;
        Headers = headers;
        Body = body;
        BodyBytes = bodyBytes;
    }

    public override string ToString() => $"#<http-request {Method} {Url}>";
}

File: scm-java/src/main/scheme/SchemeHttpRequest.java
package scheme;

import java.util.List;

public class SchemeHttpRequest {
    // Default request timeout in seconds. A value <= 0 means "no timeout".
    public static final int DEFAULT_TIMEOUT_SECONDS = 600;

    public String method;
    public String url;
    public List<String[]> headers; // each entry: [name, value]
    public String body; // null if no body — UTF-8 view of bodyBytes (for back-compat).
    public byte[] bodyBytes; // null if no body — the raw, byte-correct body.
    public int timeoutSeconds = DEFAULT_TIMEOUT_SECONDS; // <= 0 means no timeout

    public SchemeHttpRequest(String method, String url, List<String[]> headers, String body) {
        this(method, url, headers, body, null);
    }

    public SchemeHttpRequest(String method, String url, List<String[]> headers,
                             String body, byte[] bodyBytes) {
        this.method = method;
        this.url = url;
        this.headers = headers;
        this.body = body;
        this.bodyBytes = bodyBytes;
    }

    @Override
    public String toString() { return "#<http-request " + method + " " + url + ">"; }
}
